### ISSUE TITLE
[Bug] Hangs when concurrently trying to get-or-create a named actor

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#19896]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

It seems there are some bad error messages / race conditions around trying to implement getOrCreate named actor. cc @edoakes @simon-mo 

### Versions / Dependencies

master

### Reproduction script

```
import ray

@ray.remote
class Actor:
    def ping(self):
        return "ok"


@ray.remote
def getter(name):
    try:
        try:
            actor = ray.get_actor("foo")
        except:
            print("Get failed, trying to create", name)
            actor = Actor.options(name="foo", lifetime="detached").remote()
    except:
        print("Someone else created it, trying to get", name)
        actor = ray.get_actor("foo")
    return ray.get(actor.ping.remote())


def do_run(name):
    name = "actor_" + str(name)
    tasks = [getter.remote(name) for i in range(4)]
    result = ray.get(tasks)
    try:
        ray.kill(ray.get_actor(name))  # Cleanup
    except:
        pass
    return result


for i in range(1000):
    print("iteration", i, do_run(i))
```

```
2021-10-29 12:56:53,914	WARNING worker.py:1239 -- 3 retries left for task ffffffffffffffff6c821488bb8018faf2270b9701000000, attempting to resubmit.
2021-10-29 12:56:53,914	WARNING worker.py:1239 -- 3 retries left for task ffffffffffffffff4da118be0a6f36cab78fa70301000000, attempting to resubmit.
(pid=gcs_server) [2021-10-29 12:56:53,909 E 1010929 1010929] gcs_actor_manager.cc:161: Failed to register actor: Invalid: Actor with name 'foo' already exists., job id = 01000000, actor id = 6c821488bb8018faf2270b9701000000
(pid=gcs_server) [2021-10-29 12:56:53,909 E 1010929 1010929] gcs_actor_manager.cc:161: Failed to register actor: Invalid: Actor with name 'foo' already exists., job id = 01000000, actor id = 4da118be0a6f36cab78fa70301000000
(pid=gcs_server) [2021-10-29 12:56:53,909 E 1010929 1010929] gcs_actor_manager.cc:187: Failed to create actor, job id = 01000000, actor id = 6c821488bb8018faf2270b9701000000, status: Invalid: Actor may be already destroyed.
(pid=gcs_server) [2021-10-29 12:56:53,909 E 1010929 1010929] gcs_actor_manager.cc:187: Failed to create actor, job id = 01000000, actor id = 4da118be0a6f36cab78fa70301000000, status: Invalid: Actor may be already destroyed.
(getter pid=1010992) [2021-10-29 12:56:53,909 E 1010992 1011158] direct_task_transport.cc:53: Failed to create actor 4da118be0a6f36cab78fa70301000000 with status: Invalid: Actor may be already destroyed.
(getter pid=1010992) Get failed, trying to create actor_0
(getter pid=1010991) [2021-10-29 12:56:53,909 E 1010991 1011162] direct_task_transport.cc:53: Failed to create actor 6c821488bb8018faf2270b9701000000 with status: Invalid: Actor may be already destroyed.
(getter pid=1010986) Get failed, trying to create actor_0
(getter pid=1010986) Someone else created it, trying to get actor_0
(getter pid=1010991) Get failed, trying to create actor_0
(getter pid=1010994) Get failed, trying to create actor_0
```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by ericl** [comment]
(from Simon): this may be related to https://github.com/ray-project/ray/pull/17339

**Comment by jovany-wang** [comment]
Sorry for the delay.
For this issue, I don't think it's related to  #17339

Could you guys try the reproduced script on #17339  and its parent commit?

**Comment by ericl** [comment]
@jovany-wang I can confirm reverting 17339 does not fix the issue.

**Comment by rkooo567** [comment]
So I verified it worked at Ray 1.7. It is a regression from ray 1.8. I will start investigating it.

**Comment by rkooo567** [comment]
@iycheng was your async actor submission added before 1.7 right?

**Comment by iycheng** [comment]
@rkooo567  async actor submission is included in 1.8. In 1.7 it got reverted.

**Comment by rkooo567** [comment]
@iycheng turned out the issue wasn’t from that PR! So no need to worry about it!
